File: algorithms/CPlusPlus/Backtracking/n-queens.cpp
/* Code contributed by Devang Shah to MakeContributions/DSA GitHub repository

Below CPlusPlus code displays all possible solutions for a N-Queens problem.
Problem: Given a N x N chess board, arrange N queens in a way such that no two queens attack each other. Two queens are attacking each other if they are in same row, same column or diagonal
Input: N
Output: All solutions ( in grid format {Q - Queen, * - empty cell} ) of the N-queens problem
Time Complexity: O(N!)
*/

#include <iostream>
using namespace std;

int a[30];

int place(int pos)
{
    int i;
    for (i = 1; i < pos; i++)
    {
        if ((a[i] == a[pos]) || ((abs(a[i] - a[pos]) == abs(i - pos))))
            return 0;
    }
    return 1;
}

void print_sol(int n)
{
    int i, j;
    for (i = 1; i <= n; i++)
    {
        for (j = 1; j <= n; j++)
        {
            if (a[i] == j)
                cout << "Q\t";
            else
                cout << "*\t";
        }
        cout << "\n";
    }
}

void queen(int n)
{
    int k = 1;
    a[k] = 0;
    int count = 0;
    while (k != 0)
    {
        a[k] = a[k] + 1;
        while ((a[k] <= n) && !place(k))
            a[k]++;
        if (a[k] <= n)
        {
            if (k == n)
            {
                count++;
                cout << "Solution #" << count << "\n";
                print_sol(n);
            }
            else
            {
                k++;
                a[k] = 0;
            }
        }
        else
            k--;
    }
    cout << "\nTotal solutions=" << count;
}

int main()
{
    unsigned int n;
    cin >> n;
    if (n < 15)
        queen(n);
    else
        cout << "Invalid input. Program crashed";
    return 0;
}

File: algorithms/Python/backtracking/n_queens.py
n = int(input("Enter the number of queens to be placed on an nxn chessboard: "))

def isSafe(board, row, col):
    """
    Check if it's safe to place a queen at a given position on the board.

    Args:
    - board (list): The chessboard represented as a 2D list.
    - row (int): The current row being considered.
    - col (int): The current column being considered.

    Returns:
    - bool: True if it's safe to place a queen, False otherwise.
    """
    # Check the row and column
    for i in range(col):
        if board[row][i] == 1 or board[i][col] == 1:
            return False
    
    # Check upper main diagonal
    i = row
    j = col
    while i >= 0 and j >= 0:
        if board[i][j] == 1:
            return False
        i -= 1
        j -= 1
    
    # Check lower secondary diagonal
    i = row
    j = col
    while i < n and j >= 0:
        if board[i][j] == 1:
            return False
        i += 1
        j -= 1
        
    # Check main diagonal
    i = row
    j = col
    while i < n and j < n:
        if board[i][j] == 1:
            return False
        i += 1
        j += 1
    
    # Check secondary diagonal
    i = row
    j = col
    while i >= 0 and j >= 0:
        if board[i][j] == 1:
            return False
        i -= 1
        j -= 1
 
    return True

def solve(board, col):
    """
    Solve the N-Queens problem using backtracking.

    Args:
    - board (list): The chessboard represented as a 2D list.
    - col (int): The current column being considered.

    Returns:
    - bool: True if a solution is found, False otherwise.
    """
    if col >= n:
        return True
    
    for i in range(n):
        if isSafe(board, i, col):
            board[i][col] = 1
            if solve(board, col + 1):
                return True
    
        board[i][col] = 0
    return False

def printsol(s, n):
    """
    Print the solution in a readable format.

    Args:
    - s (list): The chessboard with queens placed.
    - n (int): The size of the chessboard.
    """
    for i in range(n):
        for j in range(n):
            if s[i][j] == 1:
                print('Q', end=' ')  # 'Q' represents a queen
            else:
                print('.', end=' ')  # '.' represents an empty cell
        print('\n')

def final_solve():
    """
    Initialize the chessboard and solve the N-Queens problem.
    Print the solution if one exists, otherwise, print a message.
    """
    board = [[0 for i in range(n)] for j in range(n)]
    if solve(board, 0) == False:
        print("No solution exists")
    else:
        printsol(board, n)

# Call the final_solve function to execute the solution
final_solve()